src/common/comp/knowhere_config.cc

```diff
@@ -182,4 +187,90 @@ KnowhereConfig::SetRaftMemPool() {
 #endif
 }
 
+KnowherePluginManager&
+KnowherePluginManager::Instance() {
+    static KnowherePluginManager manager;
+    return manager;
+}
+
+bool
+KnowherePluginManager::InitPlugin(std::string path) {
+    std::lock_guard<std::mutex> guard(init_mutex_);
+
+    if (inited_) {
+        return init_results_;
+    }
+
+    LOG_KNOWHERE_INFO_ << "Init plugin indexes with path = " << path;
+
+    auto splitPaths = [&](std::string input, char delimiter) {
+        std::vector<std::string> tokens;
+        std::size_t start = 0;
+        std::size_t end = input.find(delimiter);
+
+        while (end != std::string::npos) {
+            tokens.push_back(input.substr(start, end - start));
+            start = end + 1;
+            end = input.find(delimiter, start);
+        }
+
+        tokens.push_back(input.substr(start));
+        return tokens;
+    };
+
+    auto getDynamicLibs = [&](std::string path) { return splitPaths(path, ':'); };
+    auto libs = getDynamicLibs(path);
+
+    if (libs.size() == 0) {
+        LOG_KNOWHERE_INFO_ << "Plugin index list is empty";
+    }
+
+    for (auto& lib : libs) {
+        if (!std::filesystem::exists(lib)) {
+            LOG_KNOWHERE_ERROR_ << "Failed to find lib with path = " << lib << " in filesystem";
+            init_results_ = false;
+            return init_results_;
+        }
+        void* plugin = dlopen(lib.c_str(), RTLD_NOW | RTLD_GLOBAL);
+        if (!plugin) {
+            LOG_KNOWHERE_ERROR_ << "Failed to dlopen the library with path = " << lib << ", error = " << dlerror();
+            init_results_ = false;
+            inited_ = true;
+            return init_results_;
+        }
+        typedef knowhere::PluginIndexList (*GetPluginIndexes)();
+        auto getImpList = dlsym(plugin, "getPluginIndexList");
+        GetPluginIndexes getPluginIndexes = reinterpret_cast<GetPluginIndexes>(getImpList);
+        if (!getImpList) {
+            LOG_KNOWHERE_ERROR_ << "Failed to get the index list function in library with path = " << lib
+                                << ", error = " << dlerror();
+            dlclose(plugin);
+            init_results_ = false;
+            inited_ = true;
+            return init_results_;
+        }
+        auto indexList = getPluginIndexes();
```

Comment:
you're exposing C++ interface in a problematic way. It is a PITA, according to my past experience!!!
Please read https://stackoverflow.com/questions/17885060/passing-reference-to-stl-vector-over-dll-boundary , which is about MSVC, but the problems apply for linux compilers as well. Basically, when and which code allocates the `indexList` (calls `new`) and when and which code  destructs it (calls `delete`)?

I'd stick with a C interface coupled with wrappers.

Reply:
No. 
The interface should not be a C++ one, but C one.

For example,
```
// dataset C API
knowhere::CAPI_Status CreateDataset(void** dataset_handle);
knowhere::CAPI_Status GetDatasetVar(const void* dataset_handle, const char* name, void** var_data);
knowhere::CAPI_Status SetDatasetVar(const void* dataset_handle, const char* name, const void* var_data);
knowhere::CAPI_Status GetDatasetTensor(const void* dataset_handle, void** data);
knowhere::CAPI_Status GetDatasetRows(const void** dataset_handle, int64_t* rows);
knowhere::CAPI_Status DestroyDataset(const void* dataset_handle); 
...

// config C API
knowhere::CAPI_Status GetConfigInt(const void** config_handle, const char* name, int* value);
knowhere::CAPI_Status GetConfigBool(const void** config_handle, const char* name, bool* value);
...

// index C API
knowhere::Status CreateIndex(void** index_handle, const int32_t version, const knowhere::Object* object);
knowhere::Status Train(const void* index_handle, const void* dataset_handle, const void* config_handle);
knowhere::Status Search(const void* index_handle, const void* dataset_handle, const void* config_handle, void* bitset_view_data, const size_t bitset_view_size, void** result_dataset_handle);
...

```
plus a C-to-C++ wrapper that a plugin would use.

Reply:
This C interface allows to write a plugin in any language, even Python or Rust.

Reply:
Let me write a full C interface...

Reply:
#449